parts/linux/cloud-init/artifacts/ubuntu/ubuntu-snapshot-update.sh

```diff
@@ -79,11 +79,38 @@ if [ -n "${current_timestamp}" ]; then
-# upgrade from one timestamp to another timestamp
-sed -i 's/https:\/\/snapshot.ubuntu.com\/ubuntu\/\([0-9]\{8\}T[0-9]\{6\}Z\)/https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
+
+# Network isolated cluster can't access the internet, so we deploy a live patching repo service in the cluster
+# The node will use the live patching repo service to download the repo metadata and packages
+# If the annotation is not set, we will use the ubuntu snapshot repo
+live_patching_repo_service=$($KUBECTL get node ${node_name} -o jsonpath="{.metadata.annotations['kubernetes\.azure\.com/live-patching-repo-service']}")
+# Limit the live patching repo service to private IPs in the range of 10.x.x.x, 172.16.x.x - 172.31.x.x, and 192.168.x.x
+private_ip_regex="^((10\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})|(172\.(1[6-9]|2[0-9]|3[01])\.[0-9]{1,3}\.[0-9]{1,3})|(192\.168\.[0-9]{1,3}\.[0-9]{1,3}))$"
+if [[ -n "${live_patching_repo_service}" && ! "${live_patching_repo_service}" =~ $private_ip_regex ]]; then
+    echo "Ignore invalid live patching repo service: ${live_patching_repo_service}"
+    live_patching_repo_service=""
+fi
+if [ -z "${live_patching_repo_service}" ]; then
+    echo "live patching repo service is not set, use ubuntu snapshot repo"
+    # upgrade from base image to a timestamp
+    # e.g. replace https://azure.archive.ubuntu.com/ubuntu/ with https://snapshot.ubuntu.com/ubuntu/20230727T000000Z
+    sed -i 's/http:\/\/azure.archive.ubuntu.com\/ubuntu\//https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
+    # upgrade from one timestamp to another timestamp
+    # e.g. replace https://snapshot.ubuntu.com/ubuntu/20250310T000000Z with https://snapshot.ubuntu.com/ubuntu/20250318T000000Z
+    sed -i 's/https:\/\/snapshot.ubuntu.com\/ubuntu\/\([0-9]\{8\}T[0-9]\{6\}Z\)/https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
+    # No live patching repo service annotation, so we need to change to use the ubuntu snapshot repo
+    # e.g. replace http://10.224.0.5/ubuntu/ with https://snapshot.ubuntu.com/ubuntu/20250318T000000Z
+    sed -i 's/http:\/\/[0-9]\+.[0-9]\+.[0-9]\+.[0-9]\+\/ubuntu\//https:\/\/snapshot.ubuntu.com\/ubuntu\/'"${golden_timestamp}"'/g' ${source_list_path}
+else
+    echo "live patching repo service is: ${live_patching_repo_service}"
+    # upgrade from base image to live patching repo service
+    # e.g. replace https://azure.archive.ubuntu.com/ubuntu/ with http://10.224.0.5/ubuntu/
+    sed -i 's/http:\/\/azure.archive.ubuntu.com\/ubuntu\//http:\/\/'"${live_patching_repo_service}"'\/ubuntu\//g' ${source_list_path}
+    # upgrade from one ubuntu repo timestamp to live patching repo service
+    # e.g. replace https://snapshot.ubuntu.com/ubuntu/20250310T000000Z with http://10.224.0.5/ubuntu/
+    sed -i 's/https:\/\/snapshot.ubuntu.com\/ubuntu\/\([0-9]\{8\}T[0-9]\{6\}Z\)/http:\/\/'"${live_patching_repo_service}"'\/ubuntu\//g' ${source_list_path}
```

Comment:
no timestamp needed?

Reply:
can repo service ip change?

Reply:
Yes, for each patch, the live patching controller adds annotation "kubernetes.azure.com/live-patching-golden-timestamp", the annotation "kubernetes.azure.com/live-patching-repo-service" should also be added/updated. The live patching controller is responsible for making sure the value is the latest and the correct one. It will watch the service changes.

Reply:
Oh yes, you're right, let me update.

Reply:
Also need to update the non NI case to replace from repo svc ip back to snapshot svc

Reply:
Updated. Now support " replace the old ip with the new ip".

> Also need to update the non NI case to replace from repo svc ip back to snapshot svc

I think L101-L103 has covered this case. Please point it out if I misunderstand.

https://github.com/Azure/AgentBaker/blob/4918e66f9aae2de2c1d809d426defad06ebc5284/parts/linux/cloud-init/artifacts/ubuntu/ubuntu-snapshot-update.sh#L101-L103

Reply:
And L73-L76 for mariner:
https://github.com/Azure/AgentBaker/blob/4918e66f9aae2de2c1d809d426defad06ebc5284/parts/linux/cloud-init/artifacts/mariner/mariner-package-update.sh#L73-L76